File: tasks/redact-spans/python/solution/parse.py
"""Validate the redaction request and its rules."""

import re

TOP_LEVEL = {"config", "text", "rules"}
CONFIG_KEYS = {"mask", "policy", "minLength"}
LITERAL_KEYS = {"id", "kind", "value"}
SPAN_KEYS = {"id", "kind", "start", "end"}
POLICIES = {"merge", "strict"}
RULE_ID = re.compile(r"^[A-Za-z0-9_.-]+$")


def is_int(value):
    return isinstance(value, int) and not isinstance(value, bool)


def code_points(text):
    """The text as a list of Unicode code points."""
    return list(text)


def parse_document(value):
    if not isinstance(value, dict) or set(value) != TOP_LEVEL:
        raise ValueError("malformed document")
    if not isinstance(value["text"], str):
        raise ValueError("malformed text")
    if not isinstance(value["rules"], list):
        raise ValueError("malformed rules")
    return value


def parse_config(value):
    if not isinstance(value, dict) or set(value) != CONFIG_KEYS:
        raise ValueError("malformed config")
    mask = value["mask"]
    policy = value["policy"]
    minimum = value["minLength"]
    if not isinstance(mask, str) or len(code_points(mask)) != 1:
        raise ValueError("malformed mask")
    if not isinstance(policy, str) or policy not in POLICIES:
        raise ValueError("malformed policy")
    if not is_int(minimum) or minimum < 1:
        raise ValueError("malformed minimum length")
    return {"mask": mask, "policy": policy, "minLength": minimum}


def parse_rule(value, seen, length):
    if not isinstance(value, dict) or "kind" not in value:
        raise ValueError("malformed rule")
    identifier = value.get("id")
    if not isinstance(identifier, str) or not RULE_ID.match(identifier):
        raise ValueError("malformed rule id")
    if identifier in seen:
        raise ValueError("duplicate rule id")
    kind = value["kind"]
    if kind == "literal":
        if set(value) != LITERAL_KEYS:
            raise ValueError("malformed literal rule")
        text = value["value"]
        if not isinstance(text, str) or not text:
            raise ValueError("malformed literal value")
        return {"id": identifier, "kind": "literal", "value": text}
    if kind == "span":
        if set(value) != SPAN_KEYS:
            raise ValueError("malformed span rule")
        start = value["start"]
        end = value["end"]
        if not is_int(start) or not is_int(end) or start < 0:
            raise ValueError("malformed span bounds")
        if start >= end or end > length:
            raise ValueError("malformed span bounds")
        return {"id": identifier, "kind": "span", "start": start, "end": end}
    raise ValueError("unknown rule kind")

File: tasks/redact-spans/typescript/solution/parse.ts
export type Policy = "merge" | "strict";

export interface LiteralRule {
  id: string;
  kind: "literal";
  value: string;
}

export interface SpanRule {
  id: string;
  kind: "span";
  start: number;
  end: number;
}

export type Rule = LiteralRule | SpanRule;

export interface Config {
  mask: string;
  policy: Policy;
  minLength: number;
}

export interface Document {
  config: unknown;
  text: string;
  rules: unknown[];
}

export interface Span {
  start: number;
  end: number;
  rules: string[];
}

const TOP_LEVEL = ["config", "rules", "text"];
const CONFIG_KEYS = ["mask", "minLength", "policy"];
const LITERAL_KEYS = ["id", "kind", "value"];
const SPAN_KEYS = ["end", "id", "kind", "start"];
const POLICIES: Policy[] = ["merge", "strict"];
const RULE_ID = /^[A-Za-z0-9_.-]+$/;

function isObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function hasKeys(value: Record<string, unknown>, keys: string[]): boolean {
  return Object.keys(value).sort().join(",") === keys.join(",");
}

function isInteger(value: unknown): value is number {
  return typeof value === "number" && Number.isInteger(value);
}

/** The text as an array of Unicode code points. */
export function codePoints(text: string): string[] {
  return Array.from(text);
}

export function parseDocument(value: unknown): Document {
  if (!isObject(value) || !hasKeys(value, TOP_LEVEL)) {
    throw new Error("malformed document");
  }
  if (typeof value.text !== "string") {
    throw new Error("malformed text");
  }
  if (!Array.isArray(value.rules)) {
    throw new Error("malformed rules");
  }
  return { config: value.config, text: value.text, rules: value.rules };
}

export function parseConfig(value: unknown): Config {
  if (!isObject(value) || !hasKeys(value, CONFIG_KEYS)) {
    throw new Error("malformed config");
  }
  const { mask, policy, minLength } = value;
  if (typeof mask !== "string" || codePoints(mask).length !== 1) {
    throw new Error("malformed mask");
  }
  if (typeof policy !== "string" || !POLICIES.includes(policy as Policy)) {
    throw new Error("malformed policy");
  }
  if (!isInteger(minLength) || minLength < 1) {
    throw new Error("malformed minimum length");
  }
  return { mask, policy: policy as Policy, minLength };
}

export function parseRule(
  value: unknown,
  seen: Set<string>,
  length: number,
): Rule {
  if (!isObject(value) || !("kind" in value)) {
    throw new Error("malformed rule");
  }
  const id = value.id;
  if (typeof id !== "string" || !RULE_ID.test(id)) {
    throw new Error("malformed rule id");
  }
  if (seen.has(id)) {
    throw new Error("duplicate rule id");
  }
  if (value.kind === "literal") {
    if (!hasKeys(value, LITERAL_KEYS)) {
      throw new Error("malformed literal rule");
    }
    if (typeof value.value !== "string" || value.value.length === 0) {
      throw new Error("malformed literal value");
    }
    return { id, kind: "literal", value: value.value };
  }
  if (value.kind === "span") {
    if (!hasKeys(value, SPAN_KEYS)) {
      throw new Error("malformed span rule");
    }
    const { start, end } = value;
    if (!isInteger(start) || !isInteger(end) || start < 0) {
      throw new Error("malformed span bounds");
    }
    if (start >= end || end > length) {
      throw new Error("malformed span bounds");
    }
    return { id, kind: "span", start, end };
  }
  throw new Error("unknown rule kind");
}
